### PR TITLE
Support chip count change when loading grain checkpoint 

### PR DESCRIPTION
**Pull request by aireenmei**
# Description
See the added paragraph in data_input_grain.md in this PR for description of this behavior.
Also in b/452377649

# Tests
* Test scaling up:
  * Start with 1x v6e-256, run 5 steps, save checkpoints on step=0 and step=3 ([log](https://cloudlogging.app.goo.gl/nemt5wvArFi8RaGk8))
  * Resume with 2x v6e-256 with `expansion_factor_real_data=2, max_checkify=true (to make sure the PlaceHolderIterator's fake data are not used for training)`, restored the checkpoint from step=3, run until step=9 while saving checkpoints on step=6 and step=9 ([log](https://cloudlogging.app.goo.gl/4tjzLFqM4geYV3fH7))
  * Inspect the checkpoints in [base_output_directory](https://pantheon.corp.google.com/storage/browser/aireenmei-multipod/test/grain_ckpt/grain-ckpt-2025-10-28-22-52-50/checkpoints?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&e=13802955&mods=allow_workbench_image_override). For all steps, the "iter" folder contains 64 checkpoints, because it always has 64 data loading hosts. While for the "items" folder, step 0 and step 3 have 64 processes, step 6 and 9 have 128 processes.
* Test scaling down:
  * Start with 2x v6e-256, run 10 steps, save checkpoints on step 0, 3, 6, 9 ([log](https://cloudlogging.app.goo.gl/AH97Vnp1pcNUnWAf9))
  * Resume with 1x v6e-256, restore checkpoint from step=9 and run until step 15, saving checkpoints on step=12 and step=15 ([log](https://cloudlogging.app.goo.gl/MT3YCNKZL1PktyiK8))
  * Inspect the checkpoints in [base_output_directory](https://pantheon.corp.google.com/storage/browser/aireenmei-multipod/test/grain_ckpt/grain-ckpt-2025-10-28-23-20-53/checkpoints?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&e=13802955&mods=allow_workbench_image_override).  For all steps, "iter" folder contains 128 checkpoints. While for "items" folder steps 0-9 have 128 processes, step 12-15 have 64 processes.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by RissyRan** [comment]
> 🤖 Hi @aireenmei, I've received your request, and I'm working on it now! You can track my progress [in the logs](https://github.com/AI-Hypercomputer/maxtext/actions/runs/18916279794) for more details.

I see merge conflict. It's an known issue for not triggering gemini review.

File: docs/guides/data_input_grain.md [hunk 1]
```diff
@@ -70,3 +70,8 @@ eval_interval: 10000
...
 ```
+8. Experimental: resume training with different chip count
+In grain checkpoints, different data loading host has a corresponding json file. In the case that user wants to resume training with a different number of data loading host, we provide an experimental support in MaxText:
+* **Scaling up**: For example, you have a checkpoint of 64 dataloading host, and want to resume with 128 dataloading host. This is achieved by having a subset of dataloading host loading real data, then send the additional data to the other hosts. The flag `expansion_factor_real_data` has a default value of -1. When set to >1, the number of host loading real data will be `total number of host // expansion_factor_real_data`. Each data loading host loads `expansion_factor_real_data` * per_host_batch_size_to_train. For code integrity, the non-loading hosts will have a PlaceHolderDataIterator generating dummy data, which will later be discarded. User can optionally set `max_checkify=true` to enable additional checking to make sure dummy data is not used for training. In this example, set `expansion_factor_real_data=2` to go from 64 hosts to 128 hosts.
```
Comment (RissyRan) [inline docs/guides/data_input_grain.md]: nit "* per_host_batch_size_to_train" --> `* per_host_batch_size_to_train`

File: docs/guides/data_input_pipeline/data_input_grain.md [hunk 2]
```diff
@@ -70,3 +70,8 @@ eval_interval: 10000
...
+* **Scaling down**: For example, you have a checkpoint of 128 dataloading host, and want to resume with 64 dataloading host. This is achieved by restoring multiple data iterators on each host. The flag `grain_checkpoint_scaling_factor` sets the number of data iterators to be restored for each host. We then alternate between each iterator to produce a batch. In this example, set `grain_checkpoint_scaling_factor=2` to go from 128 hosts to 64 hosts.
```
Comment (RissyRan) [inline docs/guides/data_input_pipeline/data_input_grain.md]: Do you think from user perspective, could we use 1 config to do the scale up and down? 

For instance, we use `expansion_factor_real_data=2` to scale up 2x, and `expansion_factor_real_data=0.5` to scale down 2x.

File: src/MaxText/checkpointing.py [hunk 3]
```diff
@@ -44,6 +50,83 @@
...
+
+  def restore(
+      self,
+      directory: epath.Path,
+      item: Optional[grain_checkpoint_handlers.IteratorType] = None,
+      args: Any = None,
+  ) -> grain_checkpoint_handlers.IteratorType:
+    """Restores the given iterator from the checkpoint in `directory`."""
+    item = item or args.item
+    process_index = args.process_index if hasattr(args, "process_index") and args.process_index is not None else None
```
Comment (RissyRan) [inline src/MaxText/checkpointing.py]: Will this work?

```
process_index_arg = getattr(args, "process_index", None)
    process_count_arg = getattr(args, "process_count", None)
```

File: src/MaxText/checkpointing.py [hunk 4]
```diff
@@ -354,20 +439,62 @@ def map_to_pspec(data):
...
-        # Case 2: Matches if dataset type is "grain" and a specific checkpoint file exits for the iterator
-        # exists within the checkpoint manager's directory for the given step.
-        case (checkpoint_manager, dataset_type, data_iterator) if dataset_type == "grain" and data_iterator and (
-            checkpoint_manager.directory / str(step) / "iter"
-        ).exists():
-          grain_iter = grain.PyGrainCheckpointRestore(data_iterator.local_iterator)
-          return (checkpoint_manager.restore(step, args=Composite(items=checkpoint_args, iter=grain_iter)), None)
+        # Case 2: Matches if dataset type is "grain" and the data iterator is not a
+        # PlaceHolderDataIterator and a specific checkpoint file exists for the iterator
+        case (
```
Comment (RissyRan) [inline src/MaxText/checkpointing.py:540]: Sorry I am a little bit lost here. Gemini suggested some refactor using helper functions, could you please take a look and see if it fits?

```
def _prepare_scaled_up_grain_restore_args(
    self,
    data_iterator: list,
    process_count_jax: int,
    process_count_stored: int,
    directory: epath.Path
) -> MaxtextGrainCheckpointRestore:
    """
    Prepares the restore arguments for a scaled-up (list) data iterator.

    This is used when restoring a checkpoint saved with more processes than
    the current run (e.g., 64 files onto 32 JAX processes).
    """
    # 1. Validation Assertions
    assert isinstance(data_iterator, list), (
        f"{process_count_stored} processes found in Grain checkpoint {directory}, "
        f"but only {process_count_jax} JAX processes in this run. "
        "The data_iterator must be a list (e.g., grain_checkpoint_scaling_factor "
        "was set)."
    )

    scaling_factor = len(data_iterator)
    expected_process_count = process_count_stored / process_count_jax
    assert scaling_factor == expected_process_count, (
        f"Found {process_count_stored} processes in checkpoint and {process_count_jax} "
        f"JAX processes, implying a scaling factor of {expected_process_count}. "
        f"However, the data_iterator list has {scaling_factor} items."
    )

    # 2. Prepare Arguments
    local_iterator_list = [x.local_iterator for x in data_iterator]
    # Each JAX process calculates the global indices it's responsible for.
    # e.g., process 0 with scaling_factor=2 handles indices [0, 32]
    # e.g., process 1 with scaling_factor=2 handles indices [1, 33]
    process_index_list = [
        jax.process_index() + i * process_count_jax for i in range(scaling_factor)
    ]

    return MaxtextGrainCheckpointRestore(
        local_iterator_list,
        process_index=process_index_list,
        process_count=process_count_stored
    )

def _restore_grain_iterator(
    self,
    checkpoint_manager,
    step: int,
    data_iterator,
    checkpoint_args,
    expansion_factor_real_data: int  # This must be defined in the outer scope
) -> tuple[Any, None]:
    """
    Handles the complex logic for restoring a Grain data iterator checkpoint.

    This function dispatches to the correct restore strategy based on
    the number of stored checkpoint files vs. current JAX processes.
    """
    directory = checkpoint_manager.directory / str(step) / "iter"
    process_count_jax = jax.process_count()
    
    # Count the number of checkpoint files
    process_count_stored = len(list(directory.glob("process_*-of-*.json")))

    grain_restore_args = None

    if process_count_stored > process_count_jax:
        # Case 1: Scaling Up (e.g., 64 files -> 32 processes)
        grain_restore_args = self._prepare_scaled_up_grain_restore_args(
            data_iterator, process_count_jax, process_count_stored, directory
        )

    elif process_count_stored == process_count_jax:
        # Case 2: 1-to-1 Mapping (e.g., 64 files -> 64 processes)
        assert not isinstance(data_iterator, list), (
            f"{process_count_stored} processes found in checkpoint, matching "
            f"the {process_count_jax} JAX processes. "
            "Do not set grain_checkpoint_scaling_factor for 1:1 restore."
        )
        grain_restore_args = MaxtextGrainCheckpointRestore(data_iterator.local_iterator)

    elif (
        expansion_factor_real_data > 0
        and process_count_stored == process_count_jax // expansion_factor_real_data
    ):
        # Case 3: Scaling Down (e.g., 32 files -> 64 processes, with expansion)
        grain_restore_args = MaxtextGrainCheckpointRestore(
            data_iterator.local_iterator,
            process_index=jax.process_index(),
            process_count=process_count_stored
        )

    else:
        # Case 4: Mismatch
        raise ValueError(
            f"Error restoring Grain checkpoint in {directory}: "
            f"The number of stored checkpoint files ({process_count_stored}) "
            f"is incompatible with the number of JAX processes "
            f"({process_count_jax}) and data expansion "
            f"({expansion_factor_real_data})."
        )

    # Call restore once with the composed arguments
    restored_state = checkpoint_manager.restore(
        step, args=Composite(items=checkpoint_args, iter=grain_restore_args)
    )
    return (restored_state, None)


# ... (inside your `match` statement) ...
    
    # Case 2: Matches if dataset type is "grain" and a valid iterator
    # checkpoint exists. All complex logic is now in a helper method.
    case (
        checkpoint_manager,
        dataset_type,
        data_iterator,
    ) if (
        dataset_type == "grain"
        and data_iterator
        and not isinstance(data_iterator, PlaceHolderDataIterator)
        and (checkpoint_manager.directory / str(step) / "iter").exists()
    ):
        # We assume checkpoint_args and expansion_factor_real_data
        # are defined in the scope of this match statement.
        return self._restore_grain_iterator(
            checkpoint_manager,
            step,
            data_iterator,
            checkpoint_args,
            expansion_factor_real_data
        )
```

File: src/MaxText/checkpointing.py [hunk 5]
```diff
@@ -44,6 +50,83 @@
...
+        state = json.dumps(item.get_state(), indent=4)
+      else:
+        state = item.get_state().decode()
+      filename.write_text(state)
+
+    if isinstance(item, list):
+      for local_iterator, process_index, process_count in item:
+        save_single_process(local_iterator, process_index, process_count)
+    else:
+      process_index, process_count = sharding.get_process_index_and_count()
```
Comment (iindyk) [inline src/MaxText/checkpointing.py]: nit: these are just jax.process_index() and count

File: src/MaxText/checkpointing.py [hunk 6]
```diff
@@ -44,6 +50,83 @@
...
+      # https://orbax.readthedocs.io/en/latest/api_refactor.html.
+      item: Optional[grain_checkpoint_handlers.IteratorType] = None,
+      args: Any = None,
+  ):
+    """Saves the given iterator to the checkpoint in `directory`."""
+    item = item or args.item  # pytype:disable=attribute-error
+
+    def save_single_process(item, process_index, process_count):
+      filename = directory / f"process_{process_index}-of-{process_count}.json"
+      if isinstance(item, dataset.DatasetIterator):
```
Comment (iindyk) [inline src/MaxText/checkpointing.py]: this is grain.DatasetIterator

File: src/MaxText/data_loader.py [hunk 7]
```diff
@@ -35,10 +35,21 @@ class DataLoader:
...
+      self.data_iterator_index = 0
+      self.data_iterator = self.data_iterator_list[self.data_iterator_index]
+    else:
+      self.data_iterator = data_iterator
     self.last_batch = None
     self.input_data_shardings = maxtext_utils.get_input_data_sharding(config, mesh)
 
+  def update_data_iterator(self):
+    """Update to the next data iterator in the list, if applicable."""
+    if hasattr(self, "data_iterator_list"):
```
Comment (iindyk) [inline src/MaxText/data_loader.py:49]: you can make a new iterator by interleaving the list of iterators with cycle_length == len

https://google-grain.readthedocs.io/en/latest/_autosummary/grain.experimental.InterleaveIterDataset.html#grain.experimental.InterleaveIterDataset

File: src/MaxText/checkpointing.py (hunk 5 above)
Comment (aireenmei) [reply src/MaxText/checkpointing.py]: Thanks, updated

File: src/MaxText/checkpointing.py (hunk 6 above)
Comment (aireenmei) [reply src/MaxText/checkpointing.py]: Thanks, updated

File: src/MaxText/checkpointing.py (hunk 4 above)
Comment (aireenmei) [reply src/MaxText/checkpointing.py:540]: Make sense. I refactored

File: docs/guides/data_input_grain.md (hunk 1 above)
Comment (aireenmei) [reply docs/guides/data_input_grain.md]: Thanks, updated

File: src/MaxText/checkpointing.py [hunk 8]
```diff
@@ -44,6 +49,83 @@
 EmergencyReplicatorCheckpointManager = emergency_replicator_checkpoint_manager.ReplicatorCheckpointManager
 
 
+class MaxtextGrainCheckpointHandler(grain_checkpoint_handlers.CheckpointHandler):
```
Comment (SurbhiJainUSC) [inline src/MaxText/checkpointing.py]: How about we name this class as `GrainCheckpointHandler`?

File: src/MaxText/checkpointing.py [hunk 9]
```diff
@@ -44,6 +49,83 @@
...
+      return restored_items
+    else:
+      if process_index is None or process_count is None:
+        process_index, process_count = jax.process_index(), jax.process_count()
+      return restore_single_process(item, process_index, process_count)
+
+
+@ocp.args.register_with_handler(MaxtextGrainCheckpointHandler, for_save=True)
+@dataclasses.dataclass
+class MaxtextGrainCheckpointSave(ocp.args.CheckpointArgs):
```
Comment (SurbhiJainUSC) [inline src/MaxText/checkpointing.py]: nit: `GrainCheckpointSave`

File: src/MaxText/checkpointing.py [hunk 10]
```diff
@@ -44,6 +49,83 @@
...
+  item: Any
+
+
+@ocp.args.register_with_handler(MaxtextGrainCheckpointHandler, for_restore=True)
+@dataclasses.dataclass
+class MaxtextGrainCheckpointRestore(ocp.args.CheckpointArgs):
```
Comment (SurbhiJainUSC) [inline src/MaxText/checkpointing.py]: nit: `GrainCheckpointRestore`

File: src/MaxText/experimental/rl/grpo_trainer.py [hunk 11]
```diff
@@ -877,8 +877,13 @@ def generation_worker_fn(
         max_utils.print_mem_stats("After params initialized")
 
       metric_logger.buffer_and_write_train_metrics(metrics, step, step_time_delta)
-      state_to_save = _split_grpo_state(state)[0]
-      checkpointing.maybe_save_checkpoint(checkpoint_manager, state_to_save, config, data_iterator)
+
+      if config.save_checkpoint_on_completion and (config.steps - 1) % config.checkpoint_period != 0:
```
Comment (SurbhiJainUSC) [inline src/MaxText/experimental/rl/grpo_trainer.py]: Why do we need `(config.steps-1)%config.checkpoint_period!=0` here?
This condition will already be checked by `maybe_save_checkpoint`: https://github.com/AI-Hypercomputer/maxtext/blob/main/src/MaxText/checkpointing.py#L475, right?

Comment (SurbhiJainUSC) [reply src/MaxText/experimental/rl/grpo_trainer.py]: Same comment for other trainers too.

File: src/MaxText/data_loader.py (hunk 7 above)
Comment (aireenmei) [reply src/MaxText/data_loader.py:49]: Thanks for the suggestion! At this stage the data_loader has been wrapped into a MultihostDataloader object. It required some logic changes for leveraging InterleaveIterDataset. Let's leave it for now.

File: src/MaxText/checkpointing.py (hunk 8 above)
Comment (aireenmei) [reply src/MaxText/checkpointing.py]: done

File: src/MaxText/checkpointing.py (hunk 9 above)
Comment (aireenmei) [reply src/MaxText/checkpointing.py]: done

File: src/MaxText/checkpointing.py (hunk 10 above)
Comment (aireenmei) [reply src/MaxText/checkpointing.py]: done

File: src/MaxText/checkpointing.py (hunk 3 above)
Comment (aireenmei) [reply src/MaxText/checkpointing.py]: Thanks for the suggestion! Revised

File: docs/guides/data_input_pipeline/data_input_grain.md (hunk 2 above)
Comment (aireenmei) [reply docs/guides/data_input_pipeline/data_input_grain.md]: Makes sense, I thought about that, too. I revised to use only one flag.

File: src/MaxText/experimental/rl/grpo_trainer.py (hunk 11 above)
Comment (aireenmei) [reply src/MaxText/experimental/rl/grpo_trainer.py]: Thanks for the pointer. Originally I got error and partial checkpoint when the last training step is a checkpoint-saving step. I retested and indeed `(config.steps-1)%config.checkpoint_period!=0` is not necessary but it needs `checkpoint_manager.wait_until_finished()` at the end of training loop because the last checkpoint_period is not included here (https://github.com/AI-Hypercomputer/maxtext/blob/main/src/MaxText/checkpointing.py#L485-L490). I've removed  `(config.steps-1)%config.checkpoint_period!=0` for all the trainers.

**Review by RissyRan — APPROVED** [review]
